src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Semantic/DefaultRazorSemanticTokenInfoService.cs

```diff
@@ -156,11 +169,63 @@ public override void VisitMarkupTagHelperAttribute(MarkupTagHelperAttributeSynta
             {
                 if (node.TagHelperAttributeInfo.Bound)
                 {
-                    _syntaxNodes.Add(node.Name);
+                    _syntaxNodes.Add((node.Name, SyntaxKind.MarkupTagHelperAttribute));
                 }
 
                 base.VisitMarkupTagHelperAttribute(node);
             }
+
+            public override void VisitMarkupTagHelperDirectiveAttribute(MarkupTagHelperDirectiveAttributeSyntax node)
+            {
+                if (node.TagHelperAttributeInfo.Bound)
+                {
+                    _syntaxNodes.Add((node.Transition, SyntaxKind.Transition));
+                    _syntaxNodes.Add((node.Name, SyntaxKind.MarkupTagHelperDirectiveAttribute));
+
+                    if (node.Colon != null && node.ParameterName != null)
+                    {
+                        _syntaxNodes.Add((node.Colon, SyntaxKind.Colon));
+                        _syntaxNodes.Add((node.ParameterName, SyntaxKind.MarkupTagHelperDirectiveAttribute));
+                    }
+                }
+
+                base.VisitMarkupTagHelperDirectiveAttribute(node);
+            }
+
+            public override void VisitMarkupMinimizedTagHelperDirectiveAttribute(MarkupMinimizedTagHelperDirectiveAttributeSyntax node)
+            {
+                if (node.TagHelperAttributeInfo.Bound)
+                {
+                    _syntaxNodes.Add((node.Transition, SyntaxKind.Transition));
+                    _syntaxNodes.Add((node.Name, SyntaxKind.MarkupMinimizedTagHelperDirectiveAttribute));
+
+                    if (node.Colon != null && node.ParameterName != null)
+                    {
+                        _syntaxNodes.Add((node.Colon, SyntaxKind.Colon));
+                        _syntaxNodes.Add((node.ParameterName, SyntaxKind.MarkupMinimizedTagHelperDirectiveAttribute));
+                    }
+                }
+
+                base.VisitMarkupMinimizedTagHelperDirectiveAttribute(node);
+            }
+
+            private bool IsTagHelper(MarkupTagHelperElementSyntax node)
+            {
+                var name = node.StartTag.Name.Content;
+                if (IsHtmlTagName(name))
+                {
+                    var binding = node.TagHelperInfo.BindingResult;
+                    var isCaseSensitive = binding.Descriptors.All(thd => thd.CaseSensitive);
```

Comment:
RE: avoid allocating an enumerator; `Descriptors` is `IEnumerable<TagHelperDescriptor>`, which can't be indexed. I feel like given that we may as well use use the more explicit System.Linq syntax?

Reply:
Woa, shocked we made them IEnumerable. Rarely do we do that these days. As for using Linq syntax I typically avoid it because it's easy to unintentionally create a capturing lambda but in this situation I don't have too strong of a preference. Up to you